rust/src/context.rs

```diff
@@ -18,11 +18,14 @@ impl Context {
     }
 
     /// Remove all registered plugins
-    pub fn reset(&mut self) {
+    pub fn reset(&self) {
         unsafe { bindings::extism_context_reset(self.pointer) }
     }
 }
 
+unsafe impl Send for Context {}
+unsafe impl Sync for Context {}
```

Comment:
Rustler asked me to add this

Reply:
I don't think we currently don't cross thread boundaries, but if we do in the future (a multi-threaded execution env might be necessary at some point), this change means that we need to be aware of synchronizing plugin access. 

These are "marker traits" that tell the type system if any value is OK to be shared across threads. So by adding these, Rust will just take our word that there are no memory issues / data races. But, there could be. @zshipko - would be good to get your insight on this & any potential impacts, or preventative measures we should take if needed.

Reply:
Yeah, I'm not entirely sure what the consequences of adding `Send` and `Sync` here are - if we wanted to make the context struct threadsafe (which we probably do) we could wrap the plugin map in the mutex.

Reply:
I double checked and rustler just won't work without these traits. Maybe there is a way i'm missing.

```
[I] ben@m1lap ~/C/e/elixir (elixir-nif)> mix test
Compiling 1 file (.ex)
Compiling crate extism_nif in debug mode (native/extism_nif)
   Compiling extism v0.0.1-rc.3 (/Users/ben/Code/extism/rust)
   Compiling extism_nif v0.0.1-rc.3 (/Users/ben/Code/extism/elixir/native/extism_nif)
error[E0277]: `*mut extism::bindings::ExtismContext` cannot be sent between threads safely
  --> src/lib.rs:22:5
   |
22 |     rustler::resource!(ExtismContext, env);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `*mut extism::bindings::ExtismContext` cannot be sent between threads safely
   |
   = help: within `extism::Context`, the trait `Send` is not implemented for `*mut extism::bindings::ExtismContext`
   = note: required because it appears within the type `extism::Context`
   = note: required because of the requirements on the impl of `Sync` for `RwLock<extism::Context>`
note: required because it appears within the type `ExtismContext`
  --> src/lib.rs:17:8
   |
17 | struct ExtismContext {
   |        ^^^^^^^^^^^^^
note: required by a bound in `ResourceTypeProvider`
  --> /Users/ben/.cargo/registry/src/github.com-1ecc6299db9ec823/rustler-0.26.0/src/resource.rs:37:48
   |
37 | pub trait ResourceTypeProvider: Sized + Send + Sync + 'static {
   |                                                ^^^^ required by this bound in `ResourceTypeProvider`
   = note: this error originates in the macro `rustler::resource` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: `*mut extism::bindings::ExtismContext` cannot be shared between threads safely
  --> src/lib.rs:22:5
   |
22 |     rustler::resource!(ExtismContext, env);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `*mut extism::bindings::ExtismContext` cannot be shared between threads safely
   |
   = help: within `extism::Context`, the trait `Sync` is not implemented for `*mut extism::bindings::ExtismContext`
   = note: required because it appears within the type `extism::Context`
   = note: required because of the requirements on the impl of `Sync` for `RwLock<extism::Context>`
note: required because it appears within the type `ExtismContext`
  --> src/lib.rs:17:8
   |
17 | struct ExtismContext {
   |        ^^^^^^^^^^^^^
note: required by a bound in `ResourceTypeProvider`
  --> /Users/ben/.cargo/registry/src/github.com-1ecc6299db9ec823/rustler-0.26.0/src/resource.rs:37:48
   |
37 | pub trait ResourceTypeProvider: Sized + Send + Sync + 'static {
   |                                                ^^^^ required by this bound in `ResourceTypeProvider`
   = note: this error originates in the macro `rustler::resource` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0277`.
error: could not compile `extism_nif` due to 2 previous errors

== Compilation error in file lib/extism/native.ex ==
** (RuntimeError) Rust NIF compile error (rustc exit code 101)
    (rustler 0.26.0) lib/rustler/compiler.ex:41: Rustler.Compiler.compile_crate/2
    lib/extism/native.ex:2: (module)
```


I'll see if i can find a way around using the `resource!` macro. Maybe i can somehow pass the raw pointer without upsetting rust.